File: App/Resources/AlgorithmDetails/optimizedbubblesort/optimizedbubblesort.go
package main

import (
  "fmt"
)

func sort(arr []int) []int {
  i := len(arr) - 1
  for i > 0 {
    consecSorted := 1
    for j := 0; j < i; j++ {
      if arr[j] > arr[j+1] {
        arr[j], arr[j+1] = arr[j+1], arr[j]
        consecSorted = 1
      } else {
        consecSorted++
      }
    }
    i -= consecSorted
  }
  return arr
}

func main() {
  array := []int{0, 39, 21, 62, 91, 77, 14, 23,
    90, 69, 51, 81, 68, 83, 32, 56}
  fmt.Println(sort(array))
}

File: App/Resources/AlgorithmDetails/optimizedbubblesort/optimizedbubblesort.cpp
#include <cstdio>
#include <utility>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};

void printList(int arr[], int n) {
  for (int i = 0; i < n; i++) {
    if (i == 0) {
      printf("[%d, ", arr[i]);
    } else if (i != n - 1) {
      printf("%d, ", arr[i]);
    } else {
      printf("%d]", arr[i]);
    }
  }
}

void sort(int arr[], int n) {
  int i = n - 1;
  while (i > 0) {
    int consecSorted = 1;
    for (int j = 0; j < i; j++) {
      if (arr[j] > arr[j + 1]) {
        std::swap(arr[j], arr[j + 1]);
        consecSorted = 1;
      } else {
        consecSorted++;
      }
    }
    i -= consecSorted;
  }
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}
